### ISSUE TITLE
(inference-optimize is) Slower than BVLC's OpenCL(+Intel spatial)

### ISSUE THREAD

**Issue by jstumpin** [01org/caffe#63]
**Good news:**
Misclassification for batch processing (BVCL's OpenCL) has been rectified! However, prediction value remains non-probabilistic.

**Not-so-good news (as per title):**
Is it due to internal verification during initialization (see below excerpt, truncated end of the first pass out of many passes). This only happens once whenever a new network topology is introduced (e.g. 1st execution (AlexNet): verification produced, 2nd execution (AlexNet): no verification produced, 3rd execution (ResNet): verification produced). However, for AlexNet (other networks is no exception), I'm getting 8 images/sec as opposed to 110 images/sec (BVLC's OpenCL) for those consecutive executions (skipping the first one). 

> Bechmarking kernel: IDLF_float_0_1_1_64_1_1_1_1_1_1_64_64_0_0_1_64_2_2_3_1_SIMD16
>         Estimated Gflops:0.0252887
>         Estimated GFLOPS/S: 72.9681
> Bechmarking kernel: IDLF_float_0_1_1_64_1_1_1_1_1_1_64_64_0_0_1_64_2_2_2_1_SIMD16
>         Estimated Gflops:0.0252887
>         Estimated GFLOPS/S: 62.1448
> Bechmarking kernel: U_GEMM_LIKE_CONV_float_0_1_1_64_1_1_1_1_1_1_64_64_0_0_1_64_5_2_8_32_SIMD8
>         Estimated Gflops:0.0252887
>         Estimated GFLOPS/S: 249.78
> Kernel <U_GEMM_LIKE_CONV_float_0_1_1_64_1_1_1_1_1_1_64_64_0_0_1_64_5_2_8_32_SIMD8> passed verification
> Convolution Time:0.101244

**Comment by gongzg** [comment]
@jstumpin What's the meaning of "prediction value remains non-probabilistic"?

The tuning mechanism does introduce a significant tuning time for new topologies. But it's a one-time overhead for each given topology.

As to AlexNet performance issue, it's weird. Did you specify the -phase TEST parameter? Could you add both the -phase TEST and the -lt parameter when use the caffe tool to measure the time and share the output here.

**Comment by jstumpin** [comment]
Not bounded to (0,1).

Indeed it is mentioned in [wiki](https://github.com/01org/caffe/wiki/clCaffe), my bad.

You're right, it is faster (see below) but why won't this performance materialize into the classification.cpp example?

**BVCL's OpenCL**

> I0825 11:42:34.108294  9404 caffe.cpp:458] Average time per layer:
> I0825 11:42:34.109371  9404 caffe.cpp:461]      input   forward: 0.292675 ms.
> I0825 11:42:34.109371  9404 caffe.cpp:464]      input   backward: 0 ms.
> I0825 11:42:34.110374  9404 caffe.cpp:461]      conv1   forward: 1.76619 ms.
> I0825 11:42:34.110374  9404 caffe.cpp:464]      conv1   backward: 0 ms.
> I0825 11:42:34.110374  9404 caffe.cpp:461]      relu1   forward: 0.518957 ms.
> I0825 11:42:34.111377  9404 caffe.cpp:464]      relu1   backward: 0 ms.
> I0825 11:42:34.111377  9404 caffe.cpp:461]      norm1   forward: 0.604329 ms.
> I0825 11:42:34.111377  9404 caffe.cpp:464]      norm1   backward: 0 ms.
> I0825 11:42:34.112380  9404 caffe.cpp:461]      pool1   forward: 0.531115 ms.
> I0825 11:42:34.112380  9404 caffe.cpp:464]      pool1   backward: 0 ms.
> I0825 11:42:34.112380  9404 caffe.cpp:461]      conv2   forward: 1.44626 ms.
> I0825 11:42:34.113586  9404 caffe.cpp:464]      conv2   backward: 0 ms.
> I0825 11:42:34.113586  9404 caffe.cpp:461]      relu2   forward: 0.418936 ms.
> I0825 11:42:34.113586  9404 caffe.cpp:464]      relu2   backward: 0 ms.
> I0825 11:42:34.113586  9404 caffe.cpp:461]      norm2   forward: 0.659615 ms.
> I0825 11:42:34.114590  9404 caffe.cpp:464]      norm2   backward: 0 ms.
> I0825 11:42:34.114590  9404 caffe.cpp:461]      pool2   forward: 0.421144 ms.
> I0825 11:42:34.114590  9404 caffe.cpp:464]      pool2   backward: 0 ms.
> I0825 11:42:34.114590  9404 caffe.cpp:461]      conv3   forward: 1.38536 ms.
> I0825 11:42:34.115592  9404 caffe.cpp:464]      conv3   backward: 0 ms.
> I0825 11:42:34.115592  9404 caffe.cpp:461]      relu3   forward: 0.382736 ms.
> I0825 11:42:34.115592  9404 caffe.cpp:464]      relu3   backward: 0 ms.
> I0825 11:42:34.115592  9404 caffe.cpp:461]      conv4   forward: 1.48099 ms.
> I0825 11:42:34.116595  9404 caffe.cpp:464]      conv4   backward: 0 ms.
> I0825 11:42:34.116595  9404 caffe.cpp:461]      relu4   forward: 0.407217 ms.
> I0825 11:42:34.116595  9404 caffe.cpp:464]      relu4   backward: 0 ms.
> I0825 11:42:34.116595  9404 caffe.cpp:461]      conv5   forward: 1.29083 ms.
> I0825 11:42:34.117599  9404 caffe.cpp:464]      conv5   backward: 0 ms.
> I0825 11:42:34.117599  9404 caffe.cpp:461]      relu5   forward: 0.340718 ms.
> I0825 11:42:34.118603  9404 caffe.cpp:464]      relu5   backward: 0 ms.
> I0825 11:42:34.119603  9404 caffe.cpp:461]      pool5   forward: 0.40017 ms.
> I0825 11:42:34.119603  9404 caffe.cpp:464]      pool5   backward: 0 ms.
> I0825 11:42:34.119603  9404 caffe.cpp:461]        fc6   forward: 18.7198 ms.
> I0825 11:42:34.119603  9404 caffe.cpp:464]        fc6   backward: 0 ms.
> I0825 11:42:34.120606  9404 caffe.cpp:461]      relu6   forward: 0.44586 ms.
> I0825 11:42:34.120606  9404 caffe.cpp:464]      relu6   backward: 0 ms.
> I0825 11:42:34.120606  9404 caffe.cpp:461]      drop6   forward: 0.241478 ms.
> I0825 11:42:34.120606  9404 caffe.cpp:464]      drop6   backward: 0 ms.
> I0825 11:42:34.121608  9404 caffe.cpp:461]        fc7   forward: 8.62336 ms.
> I0825 11:42:34.121608  9404 caffe.cpp:464]        fc7   backward: 0 ms.
> I0825 11:42:34.121608  9404 caffe.cpp:461]      relu7   forward: 0.412659 ms.
> I0825 11:42:34.121608  9404 caffe.cpp:464]      relu7   backward: 0 ms.
> I0825 11:42:34.122611  9404 caffe.cpp:461]      drop7   forward: 0.206667 ms.
> I0825 11:42:34.122611  9404 caffe.cpp:464]      drop7   backward: 0 ms.
> I0825 11:42:34.122611  9404 caffe.cpp:461]        fc8   forward: 2.39641 ms.
> I0825 11:42:34.122611  9404 caffe.cpp:464]        fc8   backward: 0 ms.
> I0825 11:42:34.123615  9404 caffe.cpp:461]       prob   forward: 0.482518 ms.
> I0825 11:42:34.123615  9404 caffe.cpp:464]       prob   backward: 0 ms.
> I0825 11:42:34.123615  9404 caffe.cpp:469] Average Forward pass: 49.197 ms.
> I0825 11:42:34.123615  9404 caffe.cpp:471] Average Backward pass: 0 ms.
> I0825 11:42:34.124747  9404 caffe.cpp:473] Average Forward-Backward: 50.6136 ms.
> I0825 11:42:34.124747  9404 caffe.cpp:475] Total Time: 2530.68 ms.
> I0825 11:42:34.124747  9404 caffe.cpp:476] *** Benchmark ends ***

**Intel's inference-optimize**

> I0825 11:38:34.427307  7396 caffe.cpp:580] Average time per layer:
> I0825 11:38:34.428310  7396 caffe.cpp:583]      input   forward: 3.68935e+011 ms.
> I0825 11:38:34.428812  7396 caffe.cpp:586]      input   backward: 0 ms.
> I0825 11:38:34.429312  7396 caffe.cpp:583]      conv1   forward: 0.478906 ms.
> I0825 11:38:34.429813  7396 caffe.cpp:586]      conv1   backward: 0 ms.
> I0825 11:38:34.430315  7396 caffe.cpp:583]      relu1   forward: 0.207743 ms.
> I0825 11:38:34.430905  7396 caffe.cpp:586]      relu1   backward: 0 ms.
> I0825 11:38:34.431426  7396 caffe.cpp:583]      norm1   forward: 0.236942 ms.
> I0825 11:38:34.431426  7396 caffe.cpp:586]      norm1   backward: 0 ms.
> I0825 11:38:34.431910  7396 caffe.cpp:583]      pool1   forward: 0.102716 ms.
> I0825 11:38:34.431910  7396 caffe.cpp:586]      pool1   backward: 0 ms.
> I0825 11:38:34.432410  7396 caffe.cpp:583]      conv2   forward: 1.1488 ms.
> I0825 11:38:34.432912  7396 caffe.cpp:586]      conv2   backward: 0 ms.
> I0825 11:38:34.433413  7396 caffe.cpp:583]      relu2   forward: 0.174361 ms.
> I0825 11:38:34.435081  7396 caffe.cpp:586]      relu2   backward: 0 ms.
> I0825 11:38:34.435081  7396 caffe.cpp:583]      norm2   forward: 0.352791 ms.
> I0825 11:38:34.436084  7396 caffe.cpp:586]      norm2   backward: 0 ms.
> I0825 11:38:34.436585  7396 caffe.cpp:583]      pool2   forward: 0.0685908 ms.
> I0825 11:38:34.437088  7396 caffe.cpp:586]      pool2   backward: 0 ms.
> I0825 11:38:34.437590  7396 caffe.cpp:583]      conv3   forward: 0.994565 ms.
> I0825 11:38:34.438653  7396 caffe.cpp:586]      conv3   backward: 0 ms.
> I0825 11:38:34.438653  7396 caffe.cpp:583]      relu3   forward: 0.13141 ms.
> I0825 11:38:34.439649  7396 caffe.cpp:586]      relu3   backward: 0 ms.
> I0825 11:38:34.439649  7396 caffe.cpp:583]      conv4   forward: 0.971885 ms.
> I0825 11:38:34.440651  7396 caffe.cpp:586]      conv4   backward: 0 ms.
> I0825 11:38:34.440651  7396 caffe.cpp:583]      relu4   forward: 0.198907 ms.
> I0825 11:38:34.440651  7396 caffe.cpp:586]      relu4   backward: 0 ms.
> I0825 11:38:34.440651  7396 caffe.cpp:583]      conv5   forward: 1.06761 ms.
> I0825 11:38:34.441653  7396 caffe.cpp:586]      conv5   backward: 0 ms.
> I0825 11:38:34.441653  7396 caffe.cpp:583]      relu5   forward: 0.339486 ms.
> I0825 11:38:34.442656  7396 caffe.cpp:586]      relu5   backward: 0 ms.
> I0825 11:38:34.442656  7396 caffe.cpp:583]      pool5   forward: 0.0349771 ms.
> I0825 11:38:34.443658  7396 caffe.cpp:586]      pool5   backward: 0 ms.
> I0825 11:38:34.443658  7396 caffe.cpp:583]        fc6   forward: 15.9475 ms.
> I0825 11:38:34.444661  7396 caffe.cpp:586]        fc6   backward: 0 ms.
> I0825 11:38:34.445771  7396 caffe.cpp:583]      relu6   forward: 0.950123 ms.
> I0825 11:38:34.446776  7396 caffe.cpp:586]      relu6   backward: 0 ms.
> I0825 11:38:34.446776  7396 caffe.cpp:583]      drop6   forward: 0.00730682 ms.
> I0825 11:38:34.447991  7396 caffe.cpp:586]      drop6   backward: 0 ms.
> I0825 11:38:34.448357  7396 caffe.cpp:583]        fc7   forward: 6.91473 ms.
> I0825 11:38:34.448757  7396 caffe.cpp:586]        fc7   backward: 0 ms.
> I0825 11:38:34.449136  7396 caffe.cpp:583]      relu7   forward: 0.153111 ms.
> I0825 11:38:34.449506  7396 caffe.cpp:586]      relu7   backward: 0 ms.
> I0825 11:38:34.449506  7396 caffe.cpp:583]      drop7   forward: 0.0103031 ms.
> I0825 11:38:34.449506  7396 caffe.cpp:586]      drop7   backward: 0 ms.
> I0825 11:38:34.449506  7396 caffe.cpp:583]        fc8   forward: 1.75672 ms.
> I0825 11:38:34.449506  7396 caffe.cpp:586]        fc8   backward: 0 ms.
> I0825 11:38:34.450510  7396 caffe.cpp:583]       prob   forward: 0.203168 ms.
> I0825 11:38:34.450510  7396 caffe.cpp:586]       prob   backward: 0 ms.
> I0825 11:38:34.450510  7396 caffe.cpp:592] Average Forward pass: 33.4871 ms.
> I0825 11:38:34.450510  7396 caffe.cpp:594] Average Backward pass: 0 ms.
> I0825 11:38:34.451512  7396 caffe.cpp:596] Average Forward-Backward: 35.2059 ms.
> I0825 11:38:34.451512  7396 caffe.cpp:598] Total Time: 1760.3 ms.
> I0825 11:38:34.451512  7396 caffe.cpp:599] *** Benchmark ends ***

**Comment by gongzg** [comment]
@jstumpin  It seems that you haven't applied the layer fusion. You can get a little bit faster even with alexnet. As to the performance in classification.cpp, we need a one-time warm up before we start to measure the performance. The first run has some one-time overhead such as tuning or the kernel building/loading. Anyway, classification.cpp is not a deadicate benchmark case. If you want to use it to measure performance, you can change it slightly as below:

1. Do one classification for warmup.
2. In a loop to detect multiple pictures and measure the time.
3. You may also check the overhead in the preprocessing function as well. Not sure whether it is fully optimized.

The performance you got with caffe tool is only the time spent in the Forward() phase.

**Comment by gongzg** [comment]
@jstumpin As to the prediction value issue, could you share your detail instruction here then I can try to reproduce it.

**Comment by jstumpin** [comment]
Yep, saving layer fusion for later after resolving this issue. Could it be due to enabling python (didn't do so for BVLC's)? Mistakenly enabled python because at a glimpse, wiki seems to point out using python to fuse the layer. Then I realized it's just text processing, hence Caffe need not build with python.

I do use a slightly modified version of [classification.cpp](https://stackoverflow.com/questions/32668699/modifying-the-caffe-c-prediction-code-for-multiple-inputs) to process multiple images, sequentially or in batches. I've been using the same code for all test cases (Caffe's CPU-only, Caffe's OpenCL and now Intel's OpenCL) but only Intel is giving this so-called non-probabilistic output (albeit correct label). I do also taking note of several metrics (CPU usage, memory-private bytes, pre-processing pass, forward pass, prediction@total pass etc.). So do I have to modify classification.cpp differently in order to have a proper and fair evaluation or is it just python sabotaging the performance?

**Comment by gongzg** [comment]
Could you use the non-modified classification.cpp to reproduce this probabolity issue? If so, please let me know. This issue isn't related to python. I used the classification example with both fp32 and fp16, all generate correct result for the label and the probabilties.

**Comment by jstumpin** [comment]
I mean python screwing up the runtime performance, not accuracy. Here's the output for non-modified classification.cpp (same goes for BVLC's):

> classification.exe alexnet_deploy.prototxt bvlc_alexnet.caffemodel imagenet_mean.binaryproto synset_words.txt space_shuttle.jpg
> 
> Use GPU with device ID 0
> ---------- Prediction for space_shuttle.jpg ----------
> 1.0577 - "n04266014 space shuttle"
> 0.0053 - "n02690373 airliner"
> 0.0007 - "n04592741 wing"
> 0.0003 - "n04552348 warplane, military plane"
> 0.0002 - "n03773504 missile"

**Comment by gongzg** [comment]
Oh, python may introduce some overhead.
As to the classfication example, I tested it with the builtin cat image, the result is as below:

./build/examples/cpp_classification/classification.bin   models/bvlc_reference_caffenet/deploy.prototxt   models/bvlc_reference_caffenet/bvlc_reference_caffenet.caffemodel   data/ilsvrc12/imagenet_mean.binaryproto   data/ilsvrc12/synset_words.txt   examples/images/cat.jpg
Use GPU with device ID 0
---------- Prediction for examples/images/cat.jpg ----------
0.3134 - "n02123045 tabby, tabby cat"
0.2380 - "n02123159 tiger cat"
0.1235 - "n02124075 Egyptian cat"
0.1003 - "n02119022 red fox, Vulpes vulpes"
0.0715 - "n02127052 lynx, catamount"

Could you share the output of your clinfo?

**Comment by jstumpin** [comment]
Rebuilding then (without python)...

> caffe.exe device_query
> I0825 15:57:37.615366   292 common.cpp:483] Total devices: 4
> I0825 15:57:37.616539   292 common.cpp:484] CUDA devices: 0
> I0825 15:57:37.616994   292 common.cpp:485] OpenCL devices: 4
> I0825 15:57:37.617398   292 common.cpp:509] Device id:                     0
> I0825 15:57:37.617621   292 common.cpp:511] Device backend:                OpenCL
> I0825 15:57:37.617621   292 common.cpp:513] Backend details:               Intel(R) Corporation: OpenCL 2.1
> I0825 15:57:37.617621   292 common.cpp:515] Device vendor:                 Intel(R) Corporation
> I0825 15:57:37.617621   292 common.cpp:517] Name:                          Intel(R) Iris(TM) Plus Graphics 650
> I0825 15:57:37.617621   292 common.cpp:519] Total global memory:           3382260532
> I0825 15:57:37.617621   292 common.cpp:509] Device id:                     1
> I0825 15:57:37.617621   292 common.cpp:511] Device backend:                OpenCL
> I0825 15:57:37.617621   292 common.cpp:513] Backend details:               Intel(R) Corporation: OpenCL 2.1
> I0825 15:57:37.618625   292 common.cpp:515] Device vendor:                 Intel(R) Corporation
> I0825 15:57:37.618625   292 common.cpp:517] Name:                          Intel(R) Core(TM) i7-7567U CPU @ 3.50GHz
> I0825 15:57:37.618625   292 common.cpp:519] Total global memory:           8476622848
> I0825 15:57:37.618625   292 common.cpp:509] Device id:                     2
> I0825 15:57:37.619627   292 common.cpp:511] Device backend:                OpenCL
> I0825 15:57:37.619627   292 common.cpp:513] Backend details:               Intel(R) Corporation: OpenCL 2.1
> I0825 15:57:37.619627   292 common.cpp:515] Device vendor:                 Intel(R) Corporation
> I0825 15:57:37.620630   292 common.cpp:517] Name:                          Intel(R) Iris(TM) Plus Graphics 650
> I0825 15:57:37.620630   292 common.cpp:519] Total global memory:           3382260532
> I0825 15:57:37.620630   292 common.cpp:509] Device id:                     3
> I0825 15:57:37.620630   292 common.cpp:511] Device backend:                OpenCL
> I0825 15:57:37.620630   292 common.cpp:513] Backend details:               Intel(R) Corporation: OpenCL 2.1
> I0825 15:57:37.621632   292 common.cpp:515] Device vendor:                 Intel(R) Corporation
> I0825 15:57:37.621632   292 common.cpp:517] Name:                          Intel(R) Core(TM) i7-7567U CPU @ 3.50GHz
> I0825 15:57:37.621632   292 common.cpp:519] Total global memory:           8476622848

**Comment by gongzg** [comment]
The device_query is not enough, it doesn't show the driver version. Did you have clinfo installed? If not, you can use apt-get install clinfo to install it.

**Comment by jstumpin** [comment]
FYI, it's Windows.

> Number of platforms:                             2
>   Platform Profile:                              FULL_PROFILE
>   Platform Version:                              OpenCL 2.1
>   Platform Name:                                 Intel(R) OpenCL
>   Platform Vendor:                               Intel(R) Corporation
>   Platform Extensions:                           cl_intel_dx9_media_sharing cl_khr_3d_image_writes cl_khr_byte_addressable_store cl_khr_d3d11_sharing cl_khr_depth_images cl_khr_dx9_media_sharing cl_khr_fp64 cl_khr_gl_sharing cl_khr_global_int32_base_atomics cl_khr_global_int32_extended_atomics cl_khr_icd cl_khr_image2d_from_buffer cl_khr_local_int32_base_atomics cl_khr_local_int32_extended_atomics cl_khr_spir
>   Platform Profile:                              FULL_PROFILE
>   Platform Version:                              OpenCL 2.1
>   Platform Name:                                 Intel(R) OpenCL
>   Platform Vendor:                               Intel(R) Corporation
>   Platform Extensions:                           cl_intel_dx9_media_sharing cl_khr_3d_image_writes cl_khr_byte_addressable_store cl_khr_d3d11_sharing cl_khr_depth_images cl_khr_dx9_media_sharing cl_khr_fp64 cl_khr_gl_sharing cl_khr_global_int32_base_atomics cl_khr_global_int32_extended_atomics cl_khr_icd cl_khr_image2d_from_buffer cl_khr_local_int32_base_atomics cl_khr_local_int32_extended_atomics cl_khr_spir
> 
> 
>   Platform Name:                                 Intel(R) OpenCL
> Number of devices:                               2
>   Device Type:                                   CL_DEVICE_TYPE_GPU
>   Device ID:                                     32902
>   Max compute units:                             48
>   Max work items dimensions:                     3
>     Max work items[0]:                           256
>     Max work items[1]:                           256
>     Max work items[2]:                           256
>   Max work group size:                           256
>   Preferred vector width char:                   16
>   Preferred vector width short:                  8
>   Preferred vector width int:                    4
>   Preferred vector width long:                   1
>   Preferred vector width float:                  1
>   Preferred vector width double:                 1
>   Max clock frequency:                           1150Mhz
>   Address bits:                                  14757395255531667488
>   Max memory allocation:                         1561123226
>   Image support:                                 Yes
>   Max number of images read arguments:           128
>   Max number of images write arguments:          128
>   Max image 2D width:                            16384
>   Max image 2D height:                           16384
>   Max image 3D width:                            16384
>   Max image 3D height:                           16384
>   Max image 3D depth:                            2048
>   Max samplers within kernel:                    16
>   Max size of kernel argument:                   1024
>   Alignment (bits) of base address:              1024
>   Minimum alignment (bytes) for any datatype:    128
>   Single precision floating point capability
>     Denorms:                                     Yes
>     Quiet NaNs:                                  Yes
>     Round to nearest even:                       Yes
>     Round to zero:                               Yes
>     Round to +ve and infinity:                   Yes
>     IEEE754-2008 fused multiply-add:             Yes
>   Cache type:                                    Read/Write
>   Cache line size:                               64
>   Cache size:                                    1048576
>   Global memory size:                            1561123226
>   Constant buffer size:                          1561123226
>   Max number of constant args:                   8
>   Local memory type:                             Scratchpad
>   Local memory size:                             65536
>   Error correction support:                      0
>   Profiling timer resolution:                    83
>   Device endianess:                              Little
>   Available:                                     Yes
>   Compiler available:                            Yes
>   Execution capabilities:
>     Execute OpenCL kernels:                      Yes
>     Execute native function:                     No
>   Queue properties:
>     Out-of-Order:                                Yes
>     Profiling :                                  Yes
>   Platform ID:                                   014E7308
>   Name:                                          Intel(R) Iris(TM) Plus Graphics 650
>   Vendor:                                        Intel(R) Corporation
>   Driver version:                                21.20.16.4678
>   Profile:                                       FULL_PROFILE
>   Version:                                       OpenCL 2.1
>   Extensions:                                    cl_intel_accelerator cl_intel_advanced_motion_estimation cl_intel_d3d11_nv12_media_sharing cl_intel_driver_diagnostics cl_intel_dx9_media_sharing cl_intel_motion_estimation cl_intel_packed_yuv cl_intel_required_subgroup_size cl_intel_simultaneous_sharing cl_intel_subgroups cl_khr_3d_image_writes cl_khr_byte_addressable_store cl_khr_d3d10_sharing cl_khr_d3d11_sharing cl_khr_depth_images cl_khr_dx9_media_sharing cl_khr_fp16 cl_khr_fp64 cl_khr_gl_depth_images cl_khr_gl_event cl_khr_gl_msaa_sharing cl_khr_global_int32_base_atomics cl_khr_global_int32_extended_atomics cl_khr_gl_sharing cl_khr_icd cl_khr_image2d_from_buffer cl_khr_local_int32_base_atomics cl_khr_local_int32_extended_atomics cl_khr_mipmap_image cl_khr_mipmap_image_writes cl_khr_spir cl_khr_subgroups cl_khr_throttle_hints
> 
> 
>   Device Type:                                   CL_DEVICE_TYPE_CPU
>   Device ID:                                     32902
>   Max compute units:                             4
>   Max work items dimensions:                     3
>     Max work items[0]:                           8192
>     Max work items[1]:                           8192
>     Max work items[2]:                           8192
>   Max work group size:                           8192
>   Preferred vector width char:                   1
>   Preferred vector width short:                  1
>   Preferred vector width int:                    1
>   Preferred vector width long:                   1
>   Preferred vector width float:                  1
>   Preferred vector width double:                 1
>   Max clock frequency:                           3500Mhz
>   Address bits:                                  14757395255531667488
>   Max memory allocation:                         536838144
>   Image support:                                 Yes
>   Max number of images read arguments:           480
>   Max number of images write arguments:          480
>   Max image 2D width:                            16384
>   Max image 2D height:                           16384
>   Max image 3D width:                            2048
>   Max image 3D height:                           2048
>   Max image 3D depth:                            2048
>   Max samplers within kernel:                    480
>   Max size of kernel argument:                   3840
>   Alignment (bits) of base address:              1024
>   Minimum alignment (bytes) for any datatype:    128
>   Single precision floating point capability
>     Denorms:                                     Yes
>     Quiet NaNs:                                  Yes
>     Round to nearest even:                       Yes
>     Round to zero:                               No
>     Round to +ve and infinity:                   No
>     IEEE754-2008 fused multiply-add:             No
>   Cache type:                                    Read/Write
>   Cache line size:                               64
>   Cache size:                                    262144
>   Global memory size:                            536838144
>   Constant buffer size:                          131072
>   Max number of constant args:                   480
>   Local memory type:                             Global
>   Local memory size:                             32768
>   Error correction support:                      0
>   Profiling timer resolution:                    292
>   Device endianess:                              Little
>   Available:                                     Yes
>   Compiler available:                            Yes
>   Execution capabilities:
>     Execute OpenCL kernels:                      Yes
>     Execute native function:                     Yes
>   Queue properties:
>     Out-of-Order:                                Yes
>     Profiling :                                  Yes
>   Platform ID:                                   014E7308
>   Name:                                          Intel(R) Core(TM) i7-7567U CPU @ 3.50GHz
>   Vendor:                                        Intel(R) Corporation
>   Driver version:                                6.8.0.2
>   Profile:                                       FULL_PROFILE
>   Version:                                       OpenCL 2.1 (Build 2)
>   Extensions:                                    cl_khr_icd cl_khr_global_int32_base_atomics cl_khr_global_int32_extended_atomics cl_khr_local_int32_base_atomics cl_khr_local_int32_extended_atomics cl_khr_byte_addressable_store cl_khr_depth_images cl_khr_3d_image_writes cl_intel_exec_by_local_thread cl_khr_spir cl_khr_dx9_media_sharing cl_intel_dx9_media_sharing cl_khr_d3d11_sharing cl_khr_gl_sharing cl_khr_fp64 cl_khr_image2d_from_buffer
> 
> 
>   Platform Name:                                 Intel(R) OpenCL
> Number of devices:                               2
>   Device Type:                                   CL_DEVICE_TYPE_GPU
>   Device ID:                                     32902
>   Max compute units:                             48
>   Max work items dimensions:                     3
>     Max work items[0]:                           256
>     Max work items[1]:                           256
>     Max work items[2]:                           256
>   Max work group size:                           256
>   Preferred vector width char:                   16
>   Preferred vector width short:                  8
>   Preferred vector width int:                    4
>   Preferred vector width long:                   1
>   Preferred vector width float:                  1
>   Preferred vector width double:                 1
>   Max clock frequency:                           1150Mhz
>   Address bits:                                  14757395255531667488
>   Max memory allocation:                         1561123226
>   Image support:                                 Yes
>   Max number of images read arguments:           128
>   Max number of images write arguments:          128
>   Max image 2D width:                            16384
>   Max image 2D height:                           16384
>   Max image 3D width:                            16384
>   Max image 3D height:                           16384
>   Max image 3D depth:                            2048
>   Max samplers within kernel:                    16
>   Max size of kernel argument:                   1024
>   Alignment (bits) of base address:              1024
>   Minimum alignment (bytes) for any datatype:    128
>   Single precision floating point capability
>     Denorms:                                     Yes
>     Quiet NaNs:                                  Yes
>     Round to nearest even:                       Yes
>     Round to zero:                               Yes
>     Round to +ve and infinity:                   Yes
>     IEEE754-2008 fused multiply-add:             Yes
>   Cache type:                                    Read/Write
>   Cache line size:                               64
>   Cache size:                                    1048576
>   Global memory size:                            1561123226
>   Constant buffer size:                          1561123226
>   Max number of constant args:                   8
>   Local memory type:                             Scratchpad
>   Local memory size:                             65536
>   Error correction support:                      0
>   Profiling timer resolution:                    83
>   Device endianess:                              Little
>   Available:                                     Yes
>   Compiler available:                            Yes
>   Execution capabilities:
>     Execute OpenCL kernels:                      Yes
>     Execute native function:                     No
>   Queue properties:
>     Out-of-Order:                                Yes
>     Profiling :                                  Yes
>   Platform ID:                                   0153ADA8
>   Name:                                          Intel(R) Iris(TM) Plus Graphics 650
>   Vendor:                                        Intel(R) Corporation
>   Driver version:                                21.20.16.4678
>   Profile:                                       FULL_PROFILE
>   Version:                                       OpenCL 2.1
>   Extensions:                                    cl_intel_accelerator cl_intel_advanced_motion_estimation cl_intel_d3d11_nv12_media_sharing cl_intel_driver_diagnostics cl_intel_dx9_media_sharing cl_intel_motion_estimation cl_intel_packed_yuv cl_intel_required_subgroup_size cl_intel_simultaneous_sharing cl_intel_subgroups cl_khr_3d_image_writes cl_khr_byte_addressable_store cl_khr_d3d10_sharing cl_khr_d3d11_sharing cl_khr_depth_images cl_khr_dx9_media_sharing cl_khr_fp16 cl_khr_fp64 cl_khr_gl_depth_images cl_khr_gl_event cl_khr_gl_msaa_sharing cl_khr_global_int32_base_atomics cl_khr_global_int32_extended_atomics cl_khr_gl_sharing cl_khr_icd cl_khr_image2d_from_buffer cl_khr_local_int32_base_atomics cl_khr_local_int32_extended_atomics cl_khr_mipmap_image cl_khr_mipmap_image_writes cl_khr_spir cl_khr_subgroups cl_khr_throttle_hints
> 
> 
>   Device Type:                                   CL_DEVICE_TYPE_CPU
>   Device ID:                                     32902
>   Max compute units:                             4
>   Max work items dimensions:                     3
>     Max work items[0]:                           8192
>     Max work items[1]:                           8192
>     Max work items[2]:                           8192
>   Max work group size:                           8192
>   Preferred vector width char:                   1
>   Preferred vector width short:                  1
>   Preferred vector width int:                    1
>   Preferred vector width long:                   1
>   Preferred vector width float:                  1
>   Preferred vector width double:                 1
>   Max clock frequency:                           3500Mhz
>   Address bits:                                  14757395255531667488
>   Max memory allocation:                         536838144
>   Image support:                                 Yes
>   Max number of images read arguments:           480
>   Max number of images write arguments:          480
>   Max image 2D width:                            16384
>   Max image 2D height:                           16384
>   Max image 3D width:                            2048
>   Max image 3D height:                           2048
>   Max image 3D depth:                            2048
>   Max samplers within kernel:                    480
>   Max size of kernel argument:                   3840
>   Alignment (bits) of base address:              1024
>   Minimum alignment (bytes) for any datatype:    128
>   Single precision floating point capability
>     Denorms:                                     Yes
>     Quiet NaNs:                                  Yes
>     Round to nearest even:                       Yes
>     Round to zero:                               No
>     Round to +ve and infinity:                   No
>     IEEE754-2008 fused multiply-add:             No
>   Cache type:                                    Read/Write
>   Cache line size:                               64
>   Cache size:                                    262144
>   Global memory size:                            536838144
>   Constant buffer size:                          131072
>   Max number of constant args:                   480
>   Local memory type:                             Global
>   Local memory size:                             32768
>   Error correction support:                      0
>   Profiling timer resolution:                    292
>   Device endianess:                              Little
>   Available:                                     Yes
>   Compiler available:                            Yes
>   Execution capabilities:
>     Execute OpenCL kernels:                      Yes
>     Execute native function:                     Yes
>   Queue properties:
>     Out-of-Order:                                Yes
>     Profiling :                                  Yes
>   Platform ID:                                   0153ADA8
>   Name:                                          Intel(R) Core(TM) i7-7567U CPU @ 3.50GHz
>   Vendor:                                        Intel(R) Corporation
>   Driver version:                                6.8.0.2
>   Profile:                                       FULL_PROFILE
>   Version:                                       OpenCL 2.1 (Build 2)
>   Extensions:                                    cl_khr_icd cl_khr_global_int32_base_atomics cl_khr_global_int32_extended_atomics cl_khr_local_int32_base_atomics cl_khr_local_int32_extended_atomics cl_khr_byte_addressable_store cl_khr_depth_images cl_khr_3d_image_writes cl_intel_exec_by_local_thread cl_khr_spir cl_khr_dx9_media_sharing cl_intel_dx9_media_sharing cl_khr_d3d11_sharing cl_khr_gl_sharing cl_khr_fp64 cl_khr_image2d_from_buffer

**Comment by gongzg** [comment]
Your driver version : 21.20.16.4678 is a little bit out-of-date, could you try to upgrade to the latest version which is 22.20.16.4729. I just tested it on my Windows platform for the cat.jpg, I got correct output.

**Comment by jstumpin** [comment]
Updating to 22.20.16.4729 corrected the non-probabilistic output but it broke BVLC's OpenCL (the unmodified classification.cpp churns out a whole bunch of codes, as in raw c/cpp codes, and ultimately crash). 

Despite having the latest ViennaCL, it turns out I was building inference-optimize against a slightly older version of ViennaCL (the version that I built BVLC's OpenCL with). Thus I fixed the linkage part, yet the main issue persisted (if not even more slower than before). Additionally, I'm getting these during auto-tuning stage:

> warning: Linking two modules of different data layouts: '' is 'e-i64:64-v16:16-v24:32-v32:32-v48:64-v96:128-v192:256-v256:256-v512:512-v1024:1024' whereas '<origin>' is 'e-p:64:64:64-i1:8:8-i8:8:8-i16:16:16-i32:32:32-i64:64:64-f32:32:32-f64:64:64-v16:16:16-v24:32:32-v32:32:32-v48:64:64-v64:64:64-v96:128:128-v128:128:128-v192:256:256-v256:256:256-v512:512:512-v1024:1024:1024-n8:16:32:64'
> 
> warning: Linking two modules of different target triples: ' is 'spir64' whereas '<origin>' is 'vISA_64'

**Comment by gongzg** [comment]
The new driver has some issues with the BVLC OpenCL branch. I will submit patch to BVLC OpenCL branch latter after the FP16 patch got merged as I mentioned. If you want to use Caffe with Intel iGPU. I recommend you to use this inference-optimize branch for now. The warning messages are harmless you can just ignore them.

**Comment by jstumpin** [comment]
Even after re-building everything (ViennaCL & inference-optimize) using VS2017's v140 toolset (previously built using VS2013), the 1 image/sec forward pass for AlexNet is still prevalent. While warning messages are harmless, such runtime is severely prohibitive.

**Comment by gongzg** [comment]
@jstumpin 1 imgs/sec is not expected. How did you measure the performance? After upgrade the driver, could you try to run the caffe tool's device_query to check the device id and use caffe time to measure the alexnet's performance, please the following command line:

caffe time -model alexnet.prototxt -gpu igpu_device_id -phase TEST -iterations 20

Then you can run it again with -lt to get the per layer timing data and copy your output here.

**Comment by jstumpin** [comment]
Using this '[profiler](https://stackoverflow.com/questions/26677324/putting-execution-timing-code-into-a-function-opencv)' for execution time. Output for device_query is as before: 4 device IDs (0 & 2 iGPU, 1 & 3 CPU), so I'm just gonna fast forward to -lt since therein lies the problem (see bold text):

With -gpu 0 or 2:

> I0829 09:53:08.915047 11016 caffe.cpp:525] *** Benchmark begins ***
> I0829 09:53:08.915047 11016 caffe.cpp:526] Testing for 20 iterations.
> I0829 09:53:09.840988 11016 caffe.cpp:580] Average time per layer:
> I0829 09:53:09.842167 11016 caffe.cpp:583]      input   forward: 0.00373935 ms.
> I0829 09:53:09.843101 11016 caffe.cpp:586]      input   backward: 0 ms.
> I0829 09:53:09.844125 11016 caffe.cpp:583]      conv1   forward: 3.41549 ms.
> I0829 09:53:09.844125 11016 caffe.cpp:586]      conv1   backward: 0 ms.
> I0829 09:53:09.844125 11016 caffe.cpp:583]      relu1   forward: 0.804909 ms.
> I0829 09:53:09.845127 11016 caffe.cpp:586]      relu1   backward: 0 ms.
> I0829 09:53:09.845127 11016 caffe.cpp:583]      norm1   forward: 0.213059 ms.
> I0829 09:53:09.845127 11016 caffe.cpp:586]      norm1   backward: 0 ms.
> I0829 09:53:09.846129 11016 caffe.cpp:583]      pool1   forward: 0.0970778 ms.
> I0829 09:53:09.846129 11016 caffe.cpp:586]      pool1   backward: 0 ms.
> I0829 09:53:09.846129 11016 caffe.cpp:583]      conv2   forward: 6.0811 ms.
> I0829 09:53:09.846129 11016 caffe.cpp:586]      conv2   backward: 0 ms.
> I0829 09:53:09.847132 11016 caffe.cpp:583]      relu2   forward: 0.235225 ms.
> I0829 09:53:09.847132 11016 caffe.cpp:586]      relu2   backward: 0 ms.
> I0829 09:53:09.847132 11016 caffe.cpp:583]      norm2   forward: 0.331355 ms.
> I0829 09:53:09.847132 11016 caffe.cpp:586]      norm2   backward: 0 ms.
> I0829 09:53:09.847961 11016 caffe.cpp:583]      pool2   forward: 0.0774856 ms.
> I0829 09:53:09.847961 11016 caffe.cpp:586]      pool2   backward: 0 ms.
> I0829 09:53:09.847961 11016 caffe.cpp:583]      conv3   forward: 4.46845 ms.
> I0829 09:53:09.847961 11016 caffe.cpp:586]      conv3   backward: 0 ms.
> I0829 09:53:09.847961 11016 caffe.cpp:583]      relu3   forward: 0.179592 ms.
> I0829 09:53:09.848968 11016 caffe.cpp:586]      relu3   backward: 0 ms.
> I0829 09:53:09.848968 11016 caffe.cpp:583]      conv4   forward: 3.59945 ms.
> I0829 09:53:09.848968 11016 caffe.cpp:586]      conv4   backward: 0 ms.
> I0829 09:53:09.848968 11016 caffe.cpp:583]      relu4   forward: 0.255357 ms.
> I0829 09:53:09.848968 11016 caffe.cpp:586]      relu4   backward: 0 ms.
> I0829 09:53:09.848968 11016 caffe.cpp:583]      conv5   forward: 2.33806 ms.
> I0829 09:53:09.849968 11016 caffe.cpp:586]      conv5   backward: 0 ms.
> I0829 09:53:09.849968 11016 caffe.cpp:583]      relu5   forward: 0.228023 ms.
> I0829 09:53:09.849968 11016 caffe.cpp:586]      relu5   backward: 0 ms.
> I0829 09:53:09.850972 11016 caffe.cpp:583]      pool5   forward: 0.0289125 ms.
> I0829 09:53:09.850972 11016 caffe.cpp:586]      pool5   backward: 0 ms.
> I0829 09:53:09.851974 11016 caffe.cpp:583]        fc6   forward: 14.192 ms.
> I0829 09:53:09.851974 11016 caffe.cpp:586]        fc6   backward: 0 ms.
> I0829 09:53:09.853265 11016 caffe.cpp:583]      relu6   forward: 0.192161 ms.
> I0829 09:53:09.853265 11016 caffe.cpp:586]      relu6   backward: 0 ms.
> I0829 09:53:09.853265 11016 caffe.cpp:583]      drop6   forward: 0.061226 ms.
> I0829 09:53:09.854269 11016 caffe.cpp:586]      drop6   backward: 0 ms.
> I0829 09:53:09.855273 11016 caffe.cpp:583]        fc7   forward: 5.68866 ms.
> I0829 09:53:09.855273 11016 caffe.cpp:586]        fc7   backward: 0 ms.
> I0829 09:53:09.855273 11016 caffe.cpp:583]      relu7   forward: 0.212305 ms.
> I0829 09:53:09.855273 11016 caffe.cpp:586]      relu7   backward: 0 ms.
> **I0829 09:53:09.855273 11016 caffe.cpp:583]      drop7   forward: 1.84467e+12 ms.**
> I0829 09:53:09.856276 11016 caffe.cpp:586]      drop7   backward: 0 ms.
> I0829 09:53:09.856276 11016 caffe.cpp:583]        fc8   forward: 1.58223 ms.
> I0829 09:53:09.856276 11016 caffe.cpp:586]        fc8   backward: 0 ms.
> I0829 09:53:09.856276 11016 caffe.cpp:583]       prob   forward: 0.218953 ms.
> I0829 09:53:09.856276 11016 caffe.cpp:586]       prob   backward: 0 ms.
> I0829 09:53:09.857277 11016 caffe.cpp:592] Average Forward pass: 45.5077 ms.
> I0829 09:53:09.857277 11016 caffe.cpp:594] Average Backward pass: 0 ms.
> I0829 09:53:09.857277 11016 caffe.cpp:596] Average Forward-Backward: 47.0163 ms.
> I0829 09:53:09.857277 11016 caffe.cpp:598] Total Time: 940.326 ms.
> I0829 09:53:09.857277 11016 caffe.cpp:599] *** Benchmark ends ***

With -gpu 1 or 3:

> I0829 09:57:31.578793  8724 caffe.cpp:525] *** Benchmark begins ***
> I0829 09:57:31.579793  8724 caffe.cpp:526] Testing for 20 iterations.
> I0829 09:57:34.852502  8724 caffe.cpp:580] Average time per layer:
> I0829 09:57:34.852502  8724 caffe.cpp:583]      input   forward: 0.0072562 ms.
> I0829 09:57:34.854507  8724 caffe.cpp:586]      input   backward: 0 ms.
> I0829 09:57:34.854507  8724 caffe.cpp:583]      conv1   forward: 26.4121 ms.
> I0829 09:57:34.855510  8724 caffe.cpp:586]      conv1   backward: 0 ms.
> I0829 09:57:34.855510  8724 caffe.cpp:583]      relu1   forward: 0.27997 ms.
> I0829 09:57:34.855510  8724 caffe.cpp:586]      relu1   backward: 0 ms.
> I0829 09:57:34.855510  8724 caffe.cpp:583]      norm1   forward: 1.90388 ms.
> I0829 09:57:34.856513  8724 caffe.cpp:586]      norm1   backward: 0 ms.
> I0829 09:57:34.856513  8724 caffe.cpp:583]      pool1   forward: 0.422655 ms.
> I0829 09:57:34.856513  8724 caffe.cpp:586]      pool1   backward: 0 ms.
> I0829 09:57:34.857517  8724 caffe.cpp:583]      conv2   forward: 43.4261 ms.
> I0829 09:57:34.857517  8724 caffe.cpp:586]      conv2   backward: 0 ms.
> I0829 09:57:34.858538  8724 caffe.cpp:583]      relu2   forward: 0.168864 ms.
> I0829 09:57:34.858538  8724 caffe.cpp:586]      relu2   backward: 0 ms.
> I0829 09:57:34.858538  8724 caffe.cpp:583]      norm2   forward: 1.08374 ms.
> I0829 09:57:34.860523  8724 caffe.cpp:586]      norm2   backward: 0 ms.
> I0829 09:57:34.860523  8724 caffe.cpp:583]      pool2   forward: 0.224796 ms.
> I0829 09:57:34.861526  8724 caffe.cpp:586]      pool2   backward: 0 ms.
> I0829 09:57:34.861526  8724 caffe.cpp:583]      conv3   forward: 30.9255 ms.
> I0829 09:57:34.862529  8724 caffe.cpp:586]      conv3   backward: 0 ms.
> I0829 09:57:34.863533  8724 caffe.cpp:583]      relu3   forward: 0.0542536 ms.
> I0829 09:57:34.863533  8724 caffe.cpp:586]      relu3   backward: 0 ms.
> I0829 09:57:34.863533  8724 caffe.cpp:583]      conv4   forward: 24.0842 ms.
> I0829 09:57:34.864534  8724 caffe.cpp:586]      conv4   backward: 0 ms.
> I0829 09:57:34.864534  8724 caffe.cpp:583]      relu4   forward: 0.0497422 ms.
> I0829 09:57:34.864534  8724 caffe.cpp:586]      relu4   backward: 0 ms.
> I0829 09:57:34.865537  8724 caffe.cpp:583]      conv5   forward: 15.3571 ms.
> I0829 09:57:34.865537  8724 caffe.cpp:586]      conv5   backward: 0 ms.
> I0829 09:57:34.866540  8724 caffe.cpp:583]      relu5   forward: 0.0386608 ms.
> I0829 09:57:34.866540  8724 caffe.cpp:586]      relu5   backward: 0 ms.
> I0829 09:57:34.866540  8724 caffe.cpp:583]      pool5   forward: 0.0542974 ms.
> I0829 09:57:34.867543  8724 caffe.cpp:586]      pool5   backward: 0 ms.
> I0829 09:57:34.867543  8724 caffe.cpp:583]        fc6   forward: 12.1327 ms.
> I0829 09:57:34.867543  8724 caffe.cpp:586]        fc6   backward: 0 ms.
> I0829 09:57:34.868546  8724 caffe.cpp:583]      relu6   forward: 0.0347042 ms.
> I0829 09:57:34.868546  8724 caffe.cpp:586]      relu6   backward: 0 ms.
> I0829 09:57:34.869549  8724 caffe.cpp:583]      drop6   forward: 0.01095 ms.
> I0829 09:57:34.869549  8724 caffe.cpp:586]      drop6   backward: 0 ms.
> I0829 09:57:34.869549  8724 caffe.cpp:583]        fc7   forward: 4.49808 ms.
> I0829 09:57:34.869549  8724 caffe.cpp:586]        fc7   backward: 0 ms.
> I0829 09:57:34.870550  8724 caffe.cpp:583]      relu7   forward: 0.0305286 ms.
> I0829 09:57:34.871553  8724 caffe.cpp:586]      relu7   backward: 0 ms.
> **I0829 09:57:34.871553  8724 caffe.cpp:583]      drop7   forward: 0.0096798 ms.**
> I0829 09:57:34.872561  8724 caffe.cpp:586]      drop7   backward: 0 ms.
> I0829 09:57:34.872561  8724 caffe.cpp:583]        fc8   forward: 1.3875 ms.
> I0829 09:57:34.873558  8724 caffe.cpp:586]        fc8   backward: 0 ms.
> I0829 09:57:34.873558  8724 caffe.cpp:583]       prob   forward: 0.130174 ms.
> I0829 09:57:34.874560  8724 caffe.cpp:586]       prob   backward: 0 ms.
> I0829 09:57:34.874560  8724 caffe.cpp:592] Average Forward pass: 162.908 ms.
> I0829 09:57:34.874560  8724 caffe.cpp:594] Average Backward pass: 0 ms.
> I0829 09:57:34.876567  8724 caffe.cpp:596] Average Forward-Backward: 164.527 ms.
> I0829 09:57:34.876567  8724 caffe.cpp:598] Total Time: 3290.54 ms.
> I0829 09:57:34.877570  8724 caffe.cpp:599] *** Benchmark ends ***

**Comment by gongzg** [comment]
It looks like a benchmark bug however I can't reproduce it. Could you try to revert the following commit and try again:

commit c2b3a52d3c37c84702e9d5f86d7ff8d5998400be
Author: Zhigang Gong <zhigang.gong@intel.com>
Date:   Fri Aug 4 18:29:49 2017 +0800

    Refine caffe tool's time function.

    For per layer time measurement, we can use per layer timer to
    avoid per layer synchronization which could reduce most of the
    overhead due to -lt option.

    Signed-off-by: Zhigang Gong <zhigang.gong@intel.com>

**Comment by jstumpin** [comment]
You're right, that was just a bug (at Caffe's end). Also it is as you described, literally excluding the very first image for measurement solved the seemingly sluggish performance for inference-optimize. However, for ResNet50, it only yielded 7 images/sec w.r.t. BVLC's OpenCL's 11 images/sec. Will try out the layer fusion soon.

**Comment by gongzg** [comment]
That's still not totally normal.  Firstly, 01org should not be slower than BVLC. Secondly, Both 11 imges/sec or 7 images/sec ard too slow.  On my skull canyon nuc machine, I got the following performance number:

With batch size 1:
Original resnet50: 23.8 imgs/sec
Fused resnet50 : 39 imgs/sec

With batch size 16:
Original resnet50: 34.2 imgs/sec
Fused resnet50:  67.8 imgs/sec

My skull canyon is a SKL GT4e machine, your machine has a GTe3 iGPU which could get at least 3/4 of GT4e's performance. Anyway, do you get the fused version work? If so, what's the performance number?

**Comment by jstumpin** [comment]
Okay fresh start, new NUC (NUC7i7BNH, same as before, different unit). Rebuilding everything yielded the following outcome:

With batch size 1:
Original AlexNet: 1139 imgs/sec
Fused AlexNet: 947 imgs/sec
Original ResNet50: 20 imgs/sec
Fused ResNet50 : N/A

With batch size 5:
Original AlexNet: 818 imgs/sec
Fused AlexNet: 1169 imgs/sec
Original ResNet50: 27 imgs/sec
Fused ResNet50: N/A

With batch size 16:
Original AlexNet: 881 imgs/sec
Fused AlexNet: 2043 imgs/sec
Original ResNet50: 21 imgs/sec
Fused ResNet50: N/A

Built with OpenBLAS; MKL refused to tag along nicely with _model_fuse.py_. Anyway, fusing ResNet's layers led to the following (hence the N/As above): 

> I0831 00:43:05.316141  8468 layer_factory.cpp:65] Creating layer res2a_branch2c
> I0831 00:43:05.316141  8468 net.cpp:96] Creating Layer res2a_branch2c
> I0831 00:43:05.317145  8468 net.cpp:444] res2a_branch2c <- res2a_branch2b
> I0831 00:43:05.317145  8468 net.cpp:444] res2a_branch2c <- res2a_branch1
> I0831 00:43:05.317145  8468 net.cpp:413] res2a_branch2c -> res2a
> F0831 00:43:05.317145  8468 layer.hpp:430] Check failed: bottom.size() == top.size() (2 vs. 1) Convolution Layer produces one top blob as output for each bottom blob input.
> *** Check failure stack trace: ***

I reckon we can wrapped this one up and open another thread for the layer fusion error. Highly appreciate the guidance, thanks!

**Comment by gongzg** [comment]
Could you send the resnet50 you are using to me? Both the original prototxt and the fused prototxt. I can take a look.
I would like to continue to use this thread to track the original issue which is "BVLC OpenCL is faster than 01org OpenCL". For other issue, it's better to open another issue.

**Comment by jstumpin** [comment]
I'm using the original ResNet [here](https://raw.githubusercontent.com/KaimingHe/deep-residual-networks/master/prototxt/ResNet-50-deploy.prototxt). The fused version can be found [here](http://textuploader.com/d6xjj). 

Ah but thanks to you, indeed "BVLC OpenCL is faster than 01org OpenCL" has been disproved (see [reference](https://github.com/BVLC/caffe/issues/5099#issuecomment-324526350)).

**Comment by gongzg** [comment]
@jstumpin I tested the fused version and it works fine on my machine, the result is as below:

I0831 10:26:07.269515  9892 caffe.cpp:592] Average Forward pass: 25.4601 ms.
I0831 10:26:07.269539  9892 caffe.cpp:594] Average Backward pass: 0 ms.
I0831 10:26:07.269656  9892 caffe.cpp:596] Average Forward-Backward: 26.0025 ms.
I0831 10:26:07.269680  9892 caffe.cpp:598] Total Time: 1300.13 ms.
I0831 10:26:07.269683  9892 caffe.cpp:599] *** Benchmark ends ***

BTW, could you open a new issue to discuss the fused resnet problem? thanks.
